add-on/src/lib/redirect-handler/blockOrObserve.ts

```diff
@@ -0,0 +1,361 @@
+        addRules.push(generateRule(id, regexFilter, regexSubstitution))
+      }
+    }
+
+    // make sure that the default rules are added.
+    for (const { originUrl, redirectUrl } of DEFAULT_LOCAL_RULES) {
+      const { port } = new URL(state.gwURLString)
+      const regexFilter = `^${escapeURLRegex(`${originUrl}:${port}`)}(.*)$`
+      const regexSubstitution = `${redirectUrl}:${port}\\1`
+
+      if (!savedRegexFilters.has(regexFilter)) {
+        // We need to add the new rule.
+        addRules.push(saveAndGenerateRule(regexFilter, regexSubstitution))
+      }
+    }
+
+    await browser.declarativeNetRequest.updateDynamicRules({ addRules, removeRuleIds })
+  }
+}
+
+/**
+ * Saves and Generates a rule for the declarativeNetRequest API.
+ *
+ * @param regexFilter - The regex filter for the rule.
+ * @param regexSubstitution  - The regex substitution for the rule.
+ * @param excludedInitiatorDomains - The domains that are excluded from the rule.
+ * @returns
+ */
+function saveAndGenerateRule (
+  regexFilter: string,
+  regexSubstitution: string,
+  excludedInitiatorDomains: string[] = []
+): browser.DeclarativeNetRequest.Rule {
+  // We need to generate a random ID for the rule.
+  const id = Math.floor(Math.random() * 29999)
+  // We need to save the regex filter and ID to check if the rule already exists later.
+  savedRegexFilters.set(regexFilter, { id, regexSubstitution })
+  return generateRule(id, regexFilter, regexSubstitution, excludedInitiatorDomains)
+}
+
+/**
+ * Generates a rule for the declarativeNetRequest API.
+ *
+ * @param regexFilter - The regex filter for the rule.
+ * @param regexSubstitution  - The regex substitution for the rule.
+ * @param excludedInitiatorDomains - The domains that are excluded from the rule.
+ * @returns
+ */
+function generateRule (
+  id: number,
+  regexFilter: string,
+  regexSubstitution: string,
+  excludedInitiatorDomains: string[] = []
+): browser.DeclarativeNetRequest.Rule {
+  return {
+    id,
+    priority: 1,
+    action: {
+      type: 'redirect',
+      redirect: { regexSubstitution }
+    },
+    condition: {
+      regexFilter,
+      excludedInitiatorDomains,
+      resourceTypes: [
+        'csp_report',
+        'font',
+        'image',
+        'main_frame',
+        'media',
+        'object',
+        'other',
+        'ping',
+        'script',
+        'stylesheet',
+        'sub_frame',
+        'webbundle',
+        'websocket',
+        'webtransport',
+        'xmlhttprequest'
+      ]
+    }
+  }
+}
+
+/**
+ * Register a redirect rule in the dynamic rule set.
+ *
+ * @param {redirectHandlerInput} input
+ * @returns {Promise<void>}
+ */
+export function addRuleToDynamicRuleSetGenerator (
+  getState: () => CompanionState): (input: redirectHandlerInput) => Promise<void> {
+  // returning a closure to avoid passing `getState` as an argument to `addRuleToDynamicRuleSet`.
+  return async function ({ originUrl, redirectUrl }: redirectHandlerInput): Promise<void> {
```

Comment:
:warning: :thought_balloon: any particular reason why we use `originUrl` name for the first arg passed to `addRuleToDynamicRuleSet` here?



 I found it super confusing, because:
 - Origin means only domain name, port, and protocol, no path, and we use it here for passing URL with path
 -  `request.originUrl` already being a thing in the WebExtension API, we are using it in [`isCompanionRequest` origin check here](https://github.com/ipfs/ipfs-companion/blob/9f63fd850e25b3fc194cb56f8a3ba5ccdf7ac86c/add-on/src/lib/ipfs-request.js#L43-L59) and i remember it only being supported by Chromium
 
 
@whizzzkid  If you don't feel about this strongly, maybe rename it to `requestUrl` ?
[We already pass `request.url`](https://github.com/ipfs/ipfs-companion/blob/9f63fd850e25b3fc194cb56f8a3ba5ccdf7ac86c/add-on/src/lib/ipfs-request.js#L153C11-L153C34) and its less confusing name.

Reply:
I think being more clear will reduce the chance of bugs in the future, but i'm fine to defer this if there's a followup issue.